src/ShaderGen.Tests/ShaderBuiltinsTests.cs

```diff
@@ -0,0 +1,476 @@
+            Dictionary<Type, IReadOnlyCollection<FieldInfo>> childFieldInfos =
+                new Dictionary<Type, IReadOnlyCollection<FieldInfo>>();
+
+            Type currentType = typeof(T);
+            object aValue = a;
+            object bValue = b;
+            Stack<Tuple<string, Type, object, object>> stack = new Stack<Tuple<string, Type, object, object>>();
+            stack.Push(Tuple.Create(string.Empty, currentType, aValue, bValue));
+
+            while (stack.Count > 0)
+            {
+                // Pop top of stack.
+                Tuple<string, Type, object, object> tuple = stack.Pop();
+                currentType = tuple.Item2;
+                aValue = tuple.Item3;
+                bValue = tuple.Item4;
+
+                if (Equals(aValue, bValue)) continue;
+
+                // Get fields (cached)
+                if (!childFieldInfos.TryGetValue(currentType, out IReadOnlyCollection<FieldInfo> childFields))
+                    childFieldInfos.Add(currentType, childFields = currentType.GetFields().Where(f => !f.IsStatic).ToArray());
+
+                if (childFields.Count < 1)
+                {
+                    // No child fields, we have an inequality
+                    string fullName = tuple.Item1;
+                    failures.Add(Tuple.Create(fullName, aValue, bValue));
+                    continue;
+                }
+
+                foreach (FieldInfo childField in childFields)
+                {
+                    object aMemberValue = childField.GetValue(aValue);
+                    object bMemberValue = childField.GetValue(bValue);
+
+                    // Short cut equality
+                    if (Equals(aMemberValue, bMemberValue)) continue;
+
+                    string fullName = string.IsNullOrWhiteSpace(tuple.Item1)
+                        ? childField.Name
+                        : $"{tuple.Item1}.{childField.Name}";
+                    stack.Push(Tuple.Create(fullName, childField.FieldType, aMemberValue, bMemberValue));
+                }
+            }
+
+            return failures.AsReadOnly();
+        }
+
+
+        /// <summary>
+        /// The random number generators for each thread.
+        /// </summary>
+        private static readonly ThreadLocal<Random> _randomGenerators =
+            new ThreadLocal<Random>(() => new Random());
+
+        /// <summary>
+        /// Create a type with random data.
+        /// </summary>
+        /// <typeparam name="T">The random type</typeparam>
+        /// <param name="size">The optional number of bytes to fill.</param>
+        /// <returns></returns>
+        public unsafe T GetRandom<T>(int size = default(int))
+        {
+            Random random = _randomGenerators.Value;
+            size = Math.Min(Unsafe.SizeOf<T>(), size < 1 ? Int32.MaxValue : size);
+            T result = Activator.CreateInstance<T>();
+            // This buffer holds a random number
```

Comment:
Not sure what this comment is referring to

Reply:
This method originally just filled the array with random bytes in a single step, as I started mangling it to give more 'normal' floats in a limited range (to reduce the number of failures due to floating point precision) it got messy.  I had planned to replace completely but had forgotten about it.